Further reduce unsafe printf usage

geoffreygaren opened the pull request:
#### 3d5443f7963ad280d0419e45092902c3155dcd9c
<pre>
Further reduce unsafe printf usage
<a href="https://bugs.webkit.org/show_bug.cgi?id=286627">https://bugs.webkit.org/show_bug.cgi?id=286627</a>
&lt;<a href="https://rdar.apple.com/problem/143766922">rdar://problem/143766922</a>&gt;

Reviewed by Chris Dumez.

Added SAFE_FPRINTF and SAFE_SPRINTF.

Fixed up some mistakes in the way I imported the FOR_EACH macro, which trigger
when you have more than one argument.

Deployed SAFE_PRINTF, SAFE_FPRINTF, and SAFE_SPRINTF in all the simple cases.

Added a consteval ASCIILiteral constructor from string literal. This enables you
to use a string literal directly without adding _s in any context where conversion
to ASCIILiteral is unambiguous (e.g., an ASCIILIteral function argument).

(It is extremely difficult, but technically possible, to invoke the ASCIILiteral
constructor using a manually declared constant array of characters that is not
null terminated. So it includes a RELEASE_ASSERT that you didn&apos;t do that.)

* Source/JavaScriptCore/API/tests/FunctionToStringTests.cpp:
(testFunctionToString):
* Source/JavaScriptCore/API/tests/JSObjectGetProxyTargetTest.cpp:
(testJSObjectGetProxyTarget):
* Source/JavaScriptCore/API/tests/MultithreadedMultiVMExecutionTest.cpp:
(startMultithreadedMultiVMExecutionTest):
(finalizeMultithreadedMultiVMExecutionTest):
* Source/JavaScriptCore/runtime/ExceptionFuzz.cpp:
(JSC::doExceptionFuzzing):
* Source/JavaScriptCore/runtime/ExceptionFuzz.h:
(JSC::doExceptionFuzzingIfEnabled):
* Source/WTF/wtf/StdLibExtras.h:
* Source/WTF/wtf/text/ASCIILiteral.h:
* Source/WTF/wtf/text/TextStream.cpp:
(WTF::TextStream::operator&lt;&lt;):
* Source/WebCore/PAL/pal/text/TextCodec.cpp:
(PAL::TextCodec::getUnencodableReplacement):
* Source/WebCore/bindings/js/GCController.cpp:
(WebCore::GCController::dumpHeapForVM):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::showStyle):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::dumpStatistics):
(WebCore::Node::showNode const):
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::showTreeWithIndent const):
* Source/WebCore/html/HTMLElement.cpp:
(dumpInnerHTML):
* Source/WebCore/layout/layouttree/LayoutTreeBuilder.cpp:
(WebCore::Layout::printLayoutTreeForLiveDocuments):
* Source/WebCore/loader/appcache/ApplicationCache.cpp:
(WebCore::ApplicationCache::dump):
* Source/WebCore/rendering/CounterNode.cpp:
(WebCore::showTreeAndMark):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::showLayerTree):
* Source/WebKit/Shared/WebMemorySampler.cpp:
(WebKit::WebMemorySampler::stop):
(WebKit::WebMemorySampler::stopTimerFired):
* Source/WebKit/webpushd/webpushtool/WebPushToolConnection.mm:
(WebPushTool::Connection::connectToService):
(WebPushTool::Connection::sendPushMessage):
(WebPushTool::Connection::getPushPermissionState):
(WebPushTool::Connection::requestPushPermission):
* Source/WebKit/webpushd/webpushtool/WebPushToolMain.mm:
* Tools/Scripts/webkitpy/style/checkers/cpp.py:
(check_safer_cpp):

Canonical link: <a href="https://commits.webkit.org/289537@main">https://commits.webkit.org/289537@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/47198f0b79c9b0ca9129b4610882ac3f535dd2b7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87284 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6794 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41645 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92145 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/38024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89334 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7070 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14867 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/67453 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/38024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90286 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5416 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79000 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47775 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5188 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33381 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37141 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/80074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/75671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34255 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94033 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/86058 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14446 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/10513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/76260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14651 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74856 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75466 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19798 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18256 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/7378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13596 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14465 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/19758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/108552 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14210 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26122 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17653 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/15991 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/JavaScriptCore/API/tests/JSObjectGetProxyTargetTest.cpp:
```diff
@@ -42,10 +43,8 @@ int testJSObjectGetProxyTarget()
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
-        printf("    %s: %s\n", description, currentResult ? "PASS" : "FAIL");
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+    auto test = [&] (const ASCIILiteral& description, bool currentResult) {
```
We don't need `const &` for ASCIILiteral, it is a span.

cdumez commented on Source/JavaScriptCore/runtime/ExceptionFuzz.cpp:
```diff
@@ -37,7 +37,7 @@ static unsigned s_numberOfExceptionFuzzChecks;
 
 // Call this only if you know that exception fuzzing is enabled.
-void doExceptionFuzzing(JSGlobalObject* globalObject, ThrowScope& scope, const char* where, const void* returnPC)
+void doExceptionFuzzing(JSGlobalObject* globalObject, ThrowScope& scope, const ASCIILiteral& where, const void* returnPC)
```
`const &` not needed.

cdumez commented on Source/JavaScriptCore/runtime/ExceptionFuzz.h:
```diff
@@ -34,10 +35,10 @@ class JSGlobalObject;
 
 // Call this only if you know that exception fuzzing is enabled.
-void doExceptionFuzzing(JSGlobalObject*, ThrowScope&, const char* where, const void* returnPC);
+void doExceptionFuzzing(JSGlobalObject*, ThrowScope&, const ASCIILiteral& where, const void* returnPC);
```
`const &` not needed.

cdumez commented on Source/JavaScriptCore/runtime/ExceptionFuzz.h:
```diff
@@ -34,10 +35,10 @@ class JSGlobalObject;
 
 // This is what you should call if you don't know if fuzzing is enabled.
-ALWAYS_INLINE void doExceptionFuzzingIfEnabled(JSGlobalObject* globalObject, ThrowScope& scope, const char* where, const void* returnPC)
+ALWAYS_INLINE void doExceptionFuzzingIfEnabled(JSGlobalObject* globalObject, ThrowScope& scope, const ASCIILiteral& where, const void* returnPC)
```
`const &` not needed.

cdumez commented on Source/WTF/wtf/text/ASCIILiteral.h:
```diff
@@ -63,6 +63,15 @@ class ASCIILiteral final {
 
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+    template<size_t length>
+    consteval ASCIILiteral(const char (&literal)[length])
```
Why is this needed? We normally use `_s` instead.

geoffreygaren replied:
It saves typing out `_s` in the places where you pass a string literal to a function that takes an ASCIILiteral, which makes this patch smaller. 

It's type-safe and bounds-safe, so probably OK not to require `_s`?

cdumez replied:
Maybe OK as a way to reduce patch size short term but we were working towards enforcing ASCIILiteral and _s everywhere, thus my worry.

cdumez reviewed (approved):
LGTM with the fixes to pass ASCIILiteral by value.

geoffreygaren committed:
Further reduce unsafe printf usage
https://bugs.webkit.org/show_bug.cgi?id=286627
<rdar://problem/143766922>

Reviewed by Chris Dumez.

Added SAFE_FPRINTF and SAFE_SPRINTF.

Fixed up some mistakes in the way I imported the FOR_EACH macro, which trigger
when you have more than one argument.

Deployed SAFE_PRINTF, SAFE_FPRINTF, and SAFE_SPRINTF in all the simple cases.

Added a consteval ASCIILiteral constructor from string literal. This enables you
to use a string literal directly without adding _s in any context where conversion
to ASCIILiteral is unambiguous (e.g., an ASCIILIteral function argument).

(It is extremely difficult, but technically possible, to invoke the ASCIILiteral
constructor using a manually declared constant array of characters that is not
null terminated. So it includes a RELEASE_ASSERT that you didn't do that.)

* Source/JavaScriptCore/API/tests/FunctionToStringTests.cpp:
(testFunctionToString):
* Source/JavaScriptCore/API/tests/JSObjectGetProxyTargetTest.cpp:
(testJSObjectGetProxyTarget):
* Source/JavaScriptCore/API/tests/MultithreadedMultiVMExecutionTest.cpp:
(startMultithreadedMultiVMExecutionTest):
(finalizeMultithreadedMultiVMExecutionTest):
* Source/JavaScriptCore/runtime/ExceptionFuzz.cpp:
(JSC::doExceptionFuzzing):
* Source/JavaScriptCore/runtime/ExceptionFuzz.h:
(JSC::doExceptionFuzzingIfEnabled):
* Source/WTF/wtf/StdLibExtras.h:
* Source/WTF/wtf/text/ASCIILiteral.h:
* Source/WTF/wtf/text/TextStream.cpp:
(WTF::TextStream::operator<<):
* Source/WebCore/PAL/pal/text/TextCodec.cpp:
(PAL::TextCodec::getUnencodableReplacement):
* Source/WebCore/bindings/js/GCController.cpp:
(WebCore::GCController::dumpHeapForVM):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::showStyle):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::dumpStatistics):
(WebCore::Node::showNode const):
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::showTreeWithIndent const):
* Source/WebCore/html/HTMLElement.cpp:
(dumpInnerHTML):
* Source/WebCore/layout/layouttree/LayoutTreeBuilder.cpp:
(WebCore::Layout::printLayoutTreeForLiveDocuments):
* Source/WebCore/loader/appcache/ApplicationCache.cpp:
(WebCore::ApplicationCache::dump):
* Source/WebCore/rendering/CounterNode.cpp:
(WebCore::showTreeAndMark):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::showLayerTree):
* Source/WebKit/Shared/WebMemorySampler.cpp:
(WebKit::WebMemorySampler::stop):
(WebKit::WebMemorySampler::stopTimerFired):
* Source/WebKit/webpushd/webpushtool/WebPushToolConnection.mm:
(WebPushTool::Connection::connectToService):
(WebPushTool::Connection::sendPushMessage):
(WebPushTool::Connection::getPushPermissionState):
(WebPushTool::Connection::requestPushPermission):
* Source/WebKit/webpushd/webpushtool/WebPushToolMain.mm:
* Tools/Scripts/webkitpy/style/checkers/cpp.py:
(check_safer_cpp):

Canonical link: https://commits.webkit.org/289537@main
